### DISCUSSION TITLE
How to add multiple child objects to parent when updating the parent

### DISCUSSION THREAD

**Discussion by lynixiox** [prisma/prisma#23247] [Q&A] [answered]
### Question

Hi everyone. I am completely new to Typescript and Prisma. I know this is probably a super simple answer but I am trying upload a CSV of all contestants belonging to a competition but I get `Type 'CsvValus' is not assignable to type 'ContestantUncheckedUpdateManyWithoutCompetitionNestedInput | ContestantUpdateManyWithoutCompetitionNestedInput | undefined'` . How would I pass through the data and get the right type? Also what would the Prisma query look like?

### Information about Prisma Schema, Client Queries and Environment 

```prisma
model Competition {
  id    String  @id @default(cuid())
  name  String
  totalVotes Int @default(0)
  totalRevenue Int @default(0)
  startDate String
  endDate String
  contestants Contestant[]
  bundles Bundle[]
  @@unique([id])
}

model Contestant{
  id String @id @default(cuid())
  contestantId String 
  name String
  surname String
  votes Vote[]
  competition Competition @relation(fields: [contestantId], references: [id], onDelete: Cascade)
  @@index([contestantId])
  @@unique([id])
}

```

```ts

type CsvData = {
    name: string
    age: string
  }

  type CsvValus = {
    data: CsvData[]
  }

export async function updateCompetition({id,name, startDate, endDate, contestants}:{id: string, name:string, startDate: string, endDate: string, contestants: CsvValus }){

    try{
        await db.competition.update({
            where:{id},
            data: {
                name,
                startDate,
                endDate,
                contestants
            }
        })

    }catch{
        return {message: "database error: Failed ot update competition"}
    }
    revalidatePath("/dashboard/competition")
    redirect("/dashboard/")
}

```

- OS: Manjaro
- Database: PlanetScale
- Node.js version: v20.10.0
-Prisma: 5.9.1

**Comment by nurul3101** [comment]
Hello @lynixiox 👋 

The error you are seeing is because of a type mismatch between what your `updateCompetition` function is expecting for the `contestants` field and what you're attempting to pass to it. The contestants field in a Competition expects an array of Contestant records to establish a relation or update existing relations. However, your `CsvValues` type is structured as an object containing an array of a custom data type unrelated to your `Contestant` model.

Based on your model, each contestant object would need to include atleast `contestantId`, `name`, `surname`.

Your `CsvData` would need to look something like this:
```ts
// Assuming CsvData matches the Contestant model fields
type CsvData = {
    contestantId: string;
    name: string;
    surname: string;
}
```

You could then use it in your function like this:
```ts
export async function updateCompetition({
    id,
    name,
    startDate,
    endDate,
    contestants
}: {
    id: string;
    name: string;
    startDate: string;
    endDate: string;
    contestants: CsvData[] // Adjust this to directly use an array of CsvData
}) {
// Your Implementation
}
```

**Reply by lynixiox** [reply]
Hi @nurul3101 :smile: 

Thank you so much for getting back to me. I have implemented your solution as suggested however it is still throwing an error 
`Type 'CsvData[]' is not assignable to type 'ContestantUncheckedUpdateManyWithoutCompetitionNestedInput | ContestantUpdateManyWithoutCompetitionNestedInput | undefined'` 

I see that the type for contestants is defined in the `index.d.ts` file and I am unable to update this file to take the type `CsvData[]`

This is what the type looks like in `index.d.ts` 

```ts
  export type CompetitionUncheckedUpdateInput = {
    id?: StringFieldUpdateOperationsInput | string
    name?: StringFieldUpdateOperationsInput | string
    totalVotes?: IntFieldUpdateOperationsInput | number
    totalRevenue?: IntFieldUpdateOperationsInput | number
    startDate?: StringFieldUpdateOperationsInput | string
    endDate?: StringFieldUpdateOperationsInput | string
    contestants?: ContestantUncheckedUpdateManyWithoutCompetitionNestedInput
    bundles?: BundleUncheckedUpdateManyWithoutCompetitionNestedInput
  }
```

this query seems to be getting a little closer but it does not write the contestant data to the db

```ts
        await db.competition.update({
            where:{id},
            data: {
                name,
                startDate,
                endDate,
                contestants : {
                    createMany: {
                        data: contestants
                    }
                }
            }
        })
```

**Comment by lynixiox** [comment, marked as answer]
Solution 

```ts
  type CsvData = {
    name: string
    age: string
  }

  type CsvValues = {
    data: CsvData[]
  }

export async function updateCompetition(
    {   id,
        name, 
        startDate, 
        endDate, 
        contestants
    }:
    {
        id: string, 
        name:string, 
        startDate: string, 
        endDate: string, 
        contestants: CsvValues }){

    try{
        await db.competition.update({
            where:{id},
            data: {
                name,
                startDate,
                endDate,
                contestants : {
                    createMany: {
                        data: contestants.data
                    }
                }
            }
        })
```
